Source/WTF/wtf/CheckedRef.h

```diff
@@ -323,14 +323,28 @@ template <typename StorageType, typename PtrCounterType> class CanMakeCheckedPtr
                 WTFLogAlways("%s", stackTrace->toString().utf8().data());
         }
 #endif
-        // If you hit this assertion, you can turn on CHECKED_POINTER_DEBUG to help identify
-        // which CheckedPtr / CheckedRef outlived the object.
-        RELEASE_ASSERT(!m_count);
     }
 
     PtrCounterType ptrCount() const { return m_count; }
     void incrementPtrCount() const { ++m_count; }
-    void decrementPtrCount() const { ASSERT(m_count); --m_count; }
+    void decrementPtrCount() const
+    {
+        // When it detects a dangling pointer, WTF_OVERRIDE_DELETE_FOR_CHECKED_PTR scribbles an object with zeroes and then leaks it.
+        // We choose to check for zero and crash here because we don't want a dangling pointer to turn into a mysterious memory use problem.
+        // Zero is a convenient sentinel because:
+        //     * it can't appear in normal execution;
+        //     * it triggers a crash at the point where the innermost dangling CheckedPtr goes out of scope.
+        RELEASE_ASSERT(uncheckedPtrCount());
```

Comment:
Don't we also want to check this when CheckedPtr is accessed in `ptr()`?

Reply:
It's a tradeoff. I'm open to opinions / preferences here.

The argument against checking inside ptr() is:
* it's more efficient not to check, and the current CheckedPtr doesn't check
* not checking is still memory safe because we've zeroed the object, so there's no way to load a bad pointer out of it -- if you load and dereference a pointer, you'll just get a nullptr crash

(For context, Chrome has stopped paying bug bounties for UAF of zeroed+leaked objects.)

The argument for checking inside ptr() is:
* if you load a non-pointer value, you crash sooner, which can sometimes mean a clearer backtrace / bug symptom
* if you load and dereference a pointer value, you crash with a clearer assertion failure instead of a more mysterious nullptr

My understanding is that @smfr and @alanbaradlay prefer not checking because it's more efficient, so I wanted to preserve that feature for them. I wouldn't want someone to feel like they needed to create a FastCheckedPtr or something like that.

But I notice now that I can definitely add a debug ASSERT in ptr(). So I'll at least do that.

Reply:
Thanks for the clarification. It seems that we should clarify this in the commit message.

Reply:
Sounds good. I'll update it.